src/Orleans.Core/Networking/ConnectionPreambleHelper.cs

```diff
@@ -0,0 +1,87 @@
+using System;
+using System.Buffers;
+using System.Buffers.Binary;
+using System.IO.Pipelines;
+using System.Threading.Tasks;
+using Orleans.Networking.Shared;
+
+namespace Orleans.Runtime.Messaging
+{
+    internal static class ConnectionPreamble
+    {
+        private const int MaxPreambleLength = 1024;
+
+        internal static Task Write(ConnectionContext connection, GrainId grainId)
+        {
+            var output = connection.Transport.Output;
+            var grainIdByteArray = grainId.ToByteArray();
+
+            Span<byte> bytes = stackalloc byte[sizeof(int)];
+            BinaryPrimitives.WriteInt32LittleEndian(bytes, grainIdByteArray.Length);
+            var length = bytes.Length + grainIdByteArray.Length;
+
+            if (length > MaxPreambleLength)
+            {
+                throw new InvalidOperationException($"Created preamble of length {length}, which is greater than maximum allowed size of {MaxPreambleLength}.");
+            }
+
+            var buffer = output.GetSpan(length);
+            bytes.CopyTo(buffer);
+            new ReadOnlySpan<byte>(grainIdByteArray).CopyTo(buffer.Slice(sizeof(int)));
+            output.Advance(length);
+            var flushTask = output.FlushAsync();
+
+            if (flushTask.IsCompletedSuccessfully) return Task.CompletedTask;
+            return flushTask.AsTask();
+        }
+                
+        internal static async Task<GrainId> Read(ConnectionContext connection)
```

Comment:
ValueTask?

Reply:
I figure it's not a significant overhead since this occurs once per connections & connections are long-lived. I'll make the change anyway, since why not